### PR TITLE
Remove caching of INotebook 

### PR DESCRIPTION
**Pull request by DonJayamanne**
For #8160, #8143

Stop caching INotebook (found this is done in a number of places, basically the `getNotebook` has been removed completely)
Also added tests for #8143

### CODE REVIEW

**Comment by rchiodo** [comment]
Additionally the reason for the 'get' call was because in some spots we didn't actually want to create a notebook. Maybe those are irrelevant now though.

**Comment by DonJayamanne** [comment]
> Maybe those are irrelevant now though.

Agreed, couldn't find a single spot where that was used, hence removed.

**Comment by rchiodo** [comment]
> We don't use INotebookProvider directly anymore, we use IKernel.

If this is the case can we remove INotebookProvider altogether? Or it is only used to create the INotebook wrapper around the session and the kernel? (not sure what the two properties are).

Maybe it would make sense to rename INotebook to be something like ISessionContainer or something so we break cleanly from the past.

**Comment by DonJayamanne** [comment]
> n we remove INotebookProvider altogether? Or it is only used to create the INotebook wrapper around the session and the kernel? (not sure what the two properties are).

Yes thats correct.


> ebook to be something like ISessionContainer or something so we break cleanly from the past.

Sure, will do that in a separate PR (CI is super slow)

**Comment by DonJayamanne** [comment]
> ook to be something like ISessionContainer or something so we break cleanly from the past.

Not really happy with that name.
We have 2 classes in Raw case, RawJupyterSession & RawSession & then having ISessionContainer would only add to the confusion, i'd prefer to refactor this rather than adding a new name that we need to get used to again.

File: src/client/datascience/jupyter/jupyterNotebookProvider.ts [hunk 1]
```diff
@@ -30,31 +30,20 @@ export class JupyterNotebookProvider implements IJupyterNotebookProvider {
         return server?.getConnectionInfo();
     }
 
-    public async createNotebook(options: GetNotebookOptions): Promise<INotebook> {
+    public async createNotebook(options: NotebookCreationOptions): Promise<INotebook> {
         // Make sure we have a server
         const server = await this.serverProvider.getOrCreateServer({
-            getOnly: options.getOnly,
+            getOnly: false,
```
Comment (IanMatthewHuff) [inline src/client/datascience/jupyter/jupyterNotebookProvider.ts:36]: Do we use getOnly anymore? Can we just remove it from the options type?

**Comment by IanMatthewHuff** [comment]
> Additionally the reason for the 'get' call was because in some spots we didn't actually want to create a notebook. Maybe those are irrelevant now though.

I had the same thought, but I don't see the usage for that anymore. It used to be important for something like autostart.

File: src/client/datascience/jupyter/jupyterNotebookProvider.ts (hunk 1 above)
Comment (DonJayamanne) [reply src/client/datascience/jupyter/jupyterNotebookProvider.ts:36]: For Jupyter server it's required, to do with prewaring.

**Review by IanMatthewHuff — APPROVED** [review]
Code looks good to me. I think I'm just a bit hesitant since the caching always was a bit of a big ball of wax before. But no reason not to change now, we'll fix up any issues as they come up.

**Pull request merged**